add FernyTransport

allisonkarlitskaya opened the pull request:
Depends on #19 but we can start reviewing...

This is going to be the basis of the "grand unified" SSH backend (de-duplication of `cockpit.beiboot` vs `cockpit.remote`) and will probably also get used for the superuser peer.

martinpitt commented:
This needs rebasing after landing the askpass error parsing.

martinpitt commented on src/ferny/session.py:
```diff
@@ -93,8 +91,9 @@ async def connect(self,
                       interaction_responder: Optional[InteractionHandler] = None) -> None:
-        os.makedirs(FERNY_DIR, exist_ok=True)
-        self._controldir = tempfile.TemporaryDirectory(dir=FERNY_DIR)
+        rundir = os.path.join(os.environ.get('XDG_RUNTIME_DIR', '/run'), 'ferny')
```
This feels like a bad fallback -- users can't write into /run/. This should probably fall back to mkdtemp()?

allisonkarlitskaya replied:
I think the idea here was that running as a user requires `XDG_RUNTIME_DIR` to be set, and if it's not set, we assume it's root and use `/run`.  We could make this more explicit by checking the uid, but the issue you cite here is preexisting and I hesitate to change this around more than we need to — Session is on its way to less and less use, after all...

martinpitt replied:
Ah, I see now that it's moved code. Okay.

martinpitt commented on example.py:
```diff
@@ -0,0 +1,68 @@
+import argparse
```
./example.py is too generic. Should be examples/transport.py ?

allisonkarlitskaya replied:
Ya.  This is a junk commit.  It would need a bit more cleaning up anyway.

martinpitt commented on src/ferny/session.py:
```diff
@@ -141,7 +141,7 @@ async def connect(self,
             ])
 
-        agent = InteractionAgent(interaction_responder)
+        agent = InteractionAgent([interaction_responder] if interaction_responder is not None else [])
```
No veto, just a thought: Would a `*handlers` argument look nicer, instead of passing it as a list?

allisonkarlitskaya replied:
I have a strong preference against this practice.  For example, 
```python3
    subprocess.run(['ssh', 'foo'])
```

 takes its arguments as a sequence and the asyncio variants take them as a spread:

```python3
    await asyncio.create_subprocess_exec('ssh', 'foo')
```

I think the `subprocess.run()` variant is much nicer.

martinpitt commented on examples/protocol.py:
```diff
@@ -0,0 +1,78 @@
+            print('--askpass--')
+            if hint == 'none':
+                print(prompt)
+                print('[ waiting ]')
+                while True:
+                    await asyncio.sleep(10)
+                return None
```
This should get a comment. It looks like it would just sit here without *ever* actually returning from the do_askpass handler. If the function is merely to show the message (and no reply, as there is no hint), why is this necessary? This smells like a hack.

allisonkarlitskaya replied:
This is a direct copy from the other example script, and it shows more of a limitation in Python's standard library than anything else: there's no simple way to wait forever.

The intent here is definitely that we do wait forever.  The way this works at the `ssh` level is the same: `ssh-askpass` is expected to present a prompt to the user and then never exit.  `ssh` will `kill()` `ssh-askpass` when it's done with it, and we'll turn that into a cancellation of the task that the handler function is running it.

It seemed like the most direct way to map the semantics, and I'm actually pretty happy with it.  We can use context managers or `finally:` blocks here to do cleanup as we deem appropriate.  It's a bit of a moot point at present, since we don't have a reasonable way to present such prompts in the Cockpit protocol, much less a way to retract them.

martinpitt replied:
I see. It just seems odd to wait forever and then having a `return` (which should be a "not reached"?) in a handler callback. That's what I mean with "comment", expecially as the example is something that most users will look at to learn how to use it.

allisonkarlitskaya replied:
Fair enough.  I'll add a comment.

martinpitt commented on src/ferny/transport.py:
```diff
@@ -0,0 +1,371 @@
+
+
+class FernyTransport(asyncio.Transport, asyncio.SubprocessProtocol):
+    # These get initialized in connection_made(), not __init__(), but we don't
+    # list them as Optional because we won't return the object to the user
+    # until after the initialization is complete.
+    _subprocess_transport: 'asyncio.SubprocessTransport | None' = None
+    _stdin_transport: 'asyncio.WriteTransport | None' = None
+    _stdout_transport: 'asyncio.ReadTransport | None' = None
```
The comment is confusing. "We don't list them as Optional", and then the code does exactly that (well, `x | None` == `Optional[x]`)

allisonkarlitskaya replied:
This is a leftover from a simpler time.  Thanks for spotting it.

martinpitt commented on src/ferny/transport.py:
```diff
@@ -0,0 +1,371 @@
+        is_ssh: bool = True,
+        **kwargs: Any
+    ) -> 'tuple[typing.Self, P]':
+        """Connects a FernyTransport to a protocol, using the given command.
```
Please document the `args` and `is_ssh` flags here. I suppose "command" means something like `[ssh, *args]`, or does it have to include the `ssh`?

allisonkarlitskaya replied:
Sure.

martinpitt commented on src/ferny/transport.py:
```diff
@@ -0,0 +1,371 @@
+
+        The return value of this function must be held alive until at least one
+        of these things happens:
+          - it has .close() called on it
+          - .connection_lost() is called on your Protocol
```
The example awaits the `.done` future, this could be added here?

allisonkarlitskaya replied:
This is actually no longer accurate.  The condition for dropping the transport is now, unconditionally, that `connection_lost()` is called.  I'll fix that up.

allisonkarlitskaya replied:
> The example awaits the `.done` future, this could be added here?

The `.done` future is implemented inside of the example itself, from its `connection_lost()` handler:

```python
    def connection_lost(self, exc: Exception | None) -> None:
        if exc is not None:
            self.done.set_exception(exc)
        else:
            self.done.set_result(None)
```

So it's effectively doing exactly this.

martinpitt reviewed (changes requested):
Thanks! Looking forward to seeing this land soon.

allisonkarlitskaya commented on src/ferny/__init__.py:
```diff
@@ -30,3 +30,5 @@
     'write_askpass_to_tmpdir',
 ]
+
+__version__ = '0'
```
No :(

This is probably the most-requested feature of flit and it's been turned down time and time again.

This turned into a case of choosing your poison, but there were lots of reasons to avoid setuptools/setuptools_scm and this was the only reason I might have avoided flit.  Meanwhile, none of the other tools are any better....

allisonkarlitskaya replied:
Unfortunately not.  flit absolutely requires a `__version__` in your `__init__.py`.  It's one of those "opinionated" things.

martinpitt replied:
I slowly understand all these people complaining about Python's build systems..

allisonkarlitskaya commented on src/ferny/errors.py:
```diff
@@ -57,7 +57,7 @@ class ChangedHostKeyError(HostKeyError):
 # Functionality for mapping getaddrinfo()-family error messages to their
 # equivalent Python exceptions.
-def make_gaierror_map() -> Iterable[Tuple[str, int]]:
+def make_gaierror_map() -> 'Iterable[tuple[str, int]]':
```
> Stringifying the types breaks syntax highlighting in editors, so is clearly a regression.

Time for a better editor ;)

I can write more into the commit message about the benefit, but it basically boils down to "fewer `from typing import`s", "faster execution", + the ability to use the modern features without breaking older versions.

> Also, data point for the whole approach: Perhaps it's the Rust talking, but I prefer Optional[t] over t | None; your call of course

I agree that this seems a bit weird to my eye in some places, but I have a decided preference for it in others.  It also seems that this it the way the wind is blowing, even if nobody actually published anything official about it.  The first Google hit on the topic is https://stackoverflow.com/questions/69440494/python-3-10-optionaltype-or-type-none and the first answer there contains a screenshot of some youtube comment posted by a Python release manager (and black's creator) saying that `Type | None` is recommended.

Bonus points to the next poster pointing out:
> it's a very sad state of affairs if we need to fish for good practices in youtube comments :(

martinpitt replied:
Let's leave it at that.

martinpitt commented:
C8S fix:
```diff
--- .github/workflows/tests.yml
+++ .github/workflows/tests.yml
@@ -40,11 +40,12 @@ jobs:
 
       - name: Install distro runtime and test dependencies
         run: |
-          dnf install -y openssh-clients python3-pip
+          dnf install -y openssh-clients python3-pip python3-cryptography
 
       # C8S does have python3-pytest, but it's too old for us
       - name: Install pip test dependencies
-        run: pip3 install pytest pytest-asyncio asyncssh
+        # 2.14.0 and up need a too new python3-cryptography which we can't build in C8S
+        run: pip3 install pytest pytest-asyncio asyncssh==2.13.0
 
       - name: Run unit tests
         run: PYTHONPATH=src python3 -m pytest

```

This still fails three unit tests, though:
```

self = <test_basic.MockResponder object at 0x7fa6ee5e0748>, messages = ''
prompt = "Enter passphrase for key '/tmp/pytest-of-root/pytest-1/test_raise_hostkey0/keys/id_ed25519_passphrase': ", hint = ''

    async def do_askpass(self, messages: str, prompt: str, hint: str) -> 'str | None':
        # this happens on RHEL 8 which doesn't have KnownHostKey support; and everywhere with
        # handle_host_key=False, then this callback receives *every* agent interaction
        if 'fingerprint' in prompt:
            return 'yes' if await self.do_hostkey('', '', '', '', '') else 'no'
        assert 'passphrase' in prompt
        self.askpass_args.append((messages, prompt, hint))
        if isinstance(self.passphrase, Exception):
>           raise self.passphrase
E           FloatingPointError

test/test_basic.py:39: FloatingPointError
___________________________________________________ TestBasic.test_wrong_passphrase ___________________________________________________

self = <test_basic.TestBasic object at 0x7fa6ee6491d0>, key_dir = PosixPath('/tmp/pytest-of-root/pytest-1/test_wrong_passphrase0/keys')
runtime_dir = PosixPath('/tmp/pytest-of-root/pytest-1/test_wrong_passphrase0/xdg-run')

    @pytest.mark.asyncio
    async def test_wrong_passphrase(self, key_dir: pathlib.Path, runtime_dir: pathlib.Path) -> None:
        with pytest.raises(ferny.SshAuthenticationError) as raises:
            await self.run_test(key_dir, runtime_dir, True, 'xx', handle_host_key=True)
        assert 'Permission denied' in str(raises.value)
        assert 'publickey' in raises.value.methods
>       assert len(MockResponder.hostkey_args) == 1
E       assert 0 == 1
E         +0
E         -1

test/test_basic.py:226: AssertionError
__________________________________________________ TestBasic.test_correct_passphrase __________________________________________________

self = <test_basic.TestBasic object at 0x7fa6ee649668>
key_dir = PosixPath('/tmp/pytest-of-root/pytest-1/test_correct_passphrase0/keys')
runtime_dir = PosixPath('/tmp/pytest-of-root/pytest-1/test_correct_passphrase0/xdg-run')

    @pytest.mark.asyncio
    async def test_correct_passphrase(self, key_dir: pathlib.Path, runtime_dir: pathlib.Path) -> None:
        await self.run_test(key_dir, runtime_dir, True, 'passphrase', handle_host_key=True)
>       assert len(MockResponder.hostkey_args) == 1
E       assert 0 == 1
E         +0
E         -1

test/test_basic.py:235: AssertionError
```

martinpitt commented:
The three [c8s unit test failures](https://github.com/allisonkarlitskaya/ferny/actions/runs/6667311295/job/18120530305?pr=20) were relatively shallow, a regression from the "move to asyncssh" commit which en passant changed the `run_test()` default of known_host_keys from "none" to "all of them". I pushed a [separate fixup commit](https://github.com/allisonkarlitskaya/ferny/commit/34899599ee480c39d12d4cde05d3da5af2c2552c) for easy cross-review, but of course that should be squashed before landing.

That leaves the [tox eternal hang](https://github.com/allisonkarlitskaya/ferny/actions/runs/6667311293/job/18120530268), looking now.

martinpitt commented:
The tox test is rather annoying to run:
```
$ toolbox create --image ghcr.io/allisonkarlitskaya/toxbox
$ toolbox enter toxbox
Error: invalid entry point PID of container toxbox

$ podman run -it --rm -v $PWD:/src ghcr.io/allisonkarlitskaya/toxbox
mypy: install_deps> python3 -m pip install --no-index --no-build-isolation asyncssh mypy pytest-asyncio pytest-cov
ERROR: Could not find a version that satisfies the requirement asyncssh (from versions: none)
ERROR: No matching distribution found for asyncssh

mypy: exit 1 (0.84 seconds) /src> python3 -m pip install --no-index --no-build-isolation asyncssh mypy pytest-asyncio pytest-cov pid=12
mypy: FAIL ✖ in 1.06 seconds
pycodestyle: install_deps> python3 -m pip install --no-index --no-build-isolation pycodestyle
ERROR: Could not find a version that satisfies the requirement pycodestyle (from versions: none)
ERROR: No matching distribution found for pycodestyle

pycodestyle: exit 1 (0.78 seconds) /src> python3 -m pip install --no-index --no-build-isolation pycodestyle pid=24
pycodestyle: FAIL ✖ in 0.9 seconds
ruff: install_deps> python3 -m pip install --no-index --no-build-isolation ruff
ERROR: Could not find a version that satisfies the requirement ruff (from versions: none)
ERROR: No matching distribution found for ruff

ruff: exit 1 (0.77 seconds) /src> python3 -m pip install --no-index --no-build-isolation ruff pid=36
ruff: FAIL ✖ in 0.9 seconds
pytest: install_deps> python3 -m pip install --no-index --no-build-isolation asyncssh pytest-asyncio pytest-cov
ERROR: Could not find a version that satisfies the requirement asyncssh (from versions: none)
ERROR: No matching distribution found for asyncssh

pytest: exit 1 (0.79 seconds) /src> python3 -m pip install --no-index --no-build-isolation asyncssh pytest-asyncio pytest-cov pid=48
  mypy: FAIL code 1 (1.06 seconds)
  pycodestyle: FAIL code 1 (0.90 seconds)
  ruff: FAIL code 1 (0.90 seconds)
  pytest: FAIL code 1 (0.91 seconds)
  evaluation failed :( (3.80 seconds)
```

So skip the entry point:
```sh
$ podman run -it --rm -u tox -v $PWD:/src ghcr.io/allisonkarlitskaya/toxbox bash
[tox@7432606ab39a src]$ tox --colored=yes -m venv
```

This triggers a ton of failures like `PermissionError: [Errno 13] Permission denied: '/src/.coverage'` as it wants to write into the source tree, which isn't possible as `tox` user. So finally, running the container as root inside:

```
podman run -it --rm -u tox -v $PWD:/src ghcr.io/allisonkarlitskaya/toxbox bash
```
This gets rather far, but runs into some trouble with "cannot find Rust compiler" and eventually exits with 255, i.e. it tries to rebuild ruff and possibly python-crypto and such? it really shouldn't do that. 

Also, this doesn't reproduce the hang on gh workflows.

@allisonkarlitskaya how attached are you to this tox config commit? My proposal is to split it out into a separate PR, where you can hack on it on the side if you care (I *really* really don't -- you know my opinion about testing such a system integration package in a synthetic env), and we can land this PR to unblock cockpit?

martinpitt commented:
> My proposal is to split it out into a separate PR

If we do that, we of course also need to revert "workflows: stop linting using distro packages".

allisonkarlitskaya commented:
Took a middle road: kept tox, but disabled the problematic `py36-pytest` run.  Adjusted the test setup to be a bit more like dbusmock does it (a script in `test/run` downloads and runs the container now).

Fixed up another couple of silly races in the test...

It's working!

martinpitt commented on tox.ini:
```diff
@@ -0,0 +1,36 @@
+envlist = mypy,pycodestyle,ruff,pytest
+isolated_build = True
+labels =
+  venv = py312-{mypy,pycodestyle,ruff,pytest}, py3{6,7,8,9,10,11,12}-pytest
+  toxbox = py312-{mypy,pycodestyle,ruff,pytest}, py3{7,8,9,10,11,12}-pytest
```
Is the py36 difference between venv and toxbox intended? (I suppose yes, just making sure)

allisonkarlitskaya replied:
yes.  The `py36-pytest` scenario still works when I run it locally and it's my main go-to for testing, so I want to keep it there.

martinpitt reviewed (approved):
:four_leaf_clover: :rocket: Niiice!
